src/pulsar_producer.erl

```diff
@@ -61,106 +99,227 @@ callback_mode() -> [state_functions].
+                    , marshaller => fun ?MODULE:queue_item_marshaller/1
+                    , max_total_bytes => MaxTotalBytes
+                    },
+    Q = replayq:open(ReplayqCfg),
+    ProducerOpts =
+        ProducerOpts0#{ replayq => Q
+                      , max_batch_bytes => MaxBatchBytes
+                      },
     State = #state{
         partitiontopic = PartitionTopic,
-        producer_id = maps:get(producer_id, ProducerOpts),
+        producer_id = ProducerID,
         producer_name = maps:get(producer_name, ProducerOpts, pulsar_producer),
         callback = maps:get(callback, ProducerOpts, undefined),
         batch_size = maps:get(batch_size, ProducerOpts, 0),
         broker_server = BrokerServer,
         opts = pulsar_utils:maybe_enable_ssl_opts(Transport, ProducerOpts)
     },
     %% use process dict to avoid the trouble of relup
-    erlang:put(proxy_to_broker_url, ProxyToBrokerUrl),
+    put(proxy_to_broker_url, ProxyToBrokerUrl),
     {ok, idle, State, [{next_event, internal, do_connect}]}.
 
+%% idle state
+-spec idle(gen_statem:event_type(), _EventContent, state()) ->
+          gen_statem:event_handler_result(statem()).
+idle(enter, _OldState, _State) ->
+    keep_state_and_data;
 idle(_, do_connect, State) ->
     do_connect(State);
-idle({call, _From}, _Event, _State) ->
-    keep_state_and_data;
+idle({call, From}, _Event, _State) ->
+    {keep_state_and_data, [{reply, From, {error, producer_disconnected}}]};
 idle(cast, _Event, _State) ->
     {keep_state_and_data, [postpone]};
+idle(info, ?SEND_REQ(_, _) = SendRequest, State0) ->
+    State = enqueue_send_requests([SendRequest], State0),
+    {keep_state, State};
 idle(_EventType, _Event, _State) ->
     keep_state_and_data.
 
+%% connecting state
+-spec connecting(gen_statem:event_type(), _EventContent, state()) ->
+          gen_statem:event_handler_result(statem()).
+connecting(enter, _OldState, _State) ->
+    keep_state_and_data;
 connecting(_, do_connect, State) ->
     do_connect(State);
-
+connecting(info, ?SEND_REQ(_, _) = SendRequest, State0) ->
+    State = enqueue_send_requests([SendRequest], State0),
+    {keep_state, State};
+connecting(info, {CloseEvent, _Sock}, State0 = #state{})
+  when CloseEvent =:= tcp_closed; CloseEvent =:= ssl_closed ->
+    try_close_socket(State0),
+    {next_state, idle, State0#state{sock = undefined}, [{next_event, internal, do_connect}]};
 connecting(_EventType, {Inet, _, Bin}, State) when Inet == tcp; Inet == ssl ->
     {Cmd, _} = pulsar_protocol_frame:parse(Bin),
     handle_response(Cmd, State);
-
+connecting(info, Msg, _State) ->
+    logger:info("[pulsar-producer][connecting] unknown message received ~p~n  ~p", [Msg, _State]),
+    keep_state_and_data;
 connecting({call, From}, _, State) ->
-    {keep_state, State, [{reply, From ,{fail, producer_connecting}}]};
-
+    {keep_state, State, [{reply, From, {error, producer_connecting}}]};
 connecting(cast, {send, _Message}, _State) ->
```

Comment:
should "upgrade" this to a ?SEND_REQ and enqueue ?

Reply:
also add a TODO comment here to remove this clause in the next major version upgrade ?

Reply:
:heavy_check_mark:

Reply:
in theory it's possible for the new beam to receive old calls.
* process gets suspended for `advanced` upgrade
* a call is made to send the **old** `send` message, (which ended up in the mailbox)
* new version beam loaded
* process state upgraded in the `code_change` callback
* process resumed to handle messages in its mailbox

Reply:
True, good catch.  I'll add the clauses back for those specific cases, thanks!

Reply:
:heavy_check_mark: